source/server/hot_restart_impl.h

```diff
@@ -110,118 +90,39 @@ class ProcessSharedMutex : public Thread::BasicLockable {
 };
 
 /**
- * Implementation of HotRestart built for Linux.
+ * Implementation of HotRestart built for Linux. Most of the "protocol" type logic is split out into
+ * HotRestarting{Base,Parent,Child}. This class ties all that to shared memory and version logic.
  */
-class HotRestartImpl : public HotRestart, Logger::Loggable<Logger::Id::main> {
+class HotRestartImpl : public HotRestart {
 public:
-  HotRestartImpl(const Options& options, Stats::SymbolTable& symbol_table);
+  HotRestartImpl(const Options& options);
 
   // Server::HotRestart
   void drainParentListeners() override;
   int duplicateParentListenSocket(const std::string& address) override;
-  void getParentStats(GetParentStatsInfo& info) override;
   void initialize(Event::Dispatcher& dispatcher, Server::Instance& server) override;
-  void shutdownParentAdmin(ShutdownParentAdminInfo& info) override;
-  void terminateParent() override;
+  void sendParentAdminShutdownRequest(time_t& original_start_time) override;
```

Comment:
This value is passed around in protobufs, and only used as a stat. The use of protobufs would make, if anything, the protobuf Timestamp type the right choice. However, since it's only dumped into a stat, I'm thinking it's best left alone.

Reply:
If it's main representation is as a stat then I think leaving it as a uint64_t makes sense, and then give it units (_sec) wherever it's carried like that.

When you get the actual time from the system we should use systemTime() I think.

I don't think time_t is the right choice for protobufs or stats, and we have a more appropriate time abstraction (which can be mocked, for example) to use. So I still think getting rid of time_t is appropriate but we can carry it around as `uint64_t original_start_time_sec`

Reply:
Envoy code may generate this value with a time() call, but it only *uses* values of it (by simply writing them into a plain uint64 stat) that have been gotten by RPC from another Envoy process, so that's the place where you would choose your values. Maybe if there was some testing that cared about what the time values were in the RPC messages you were sending, based on the behavior and timing of the rest of the code? But, this is something that should be recorded just once, right at startup. I will definitely add `_sec`, though; when it's 64 bits you can never be sure between seconds and ms.